### PR TITLE
Improve R Package Dependency Resolution and Script Execution Time

### PR DESCRIPTION
**Pull request by seantleonard**
# Why is this change being made?

Mentioned in #95, script execution time can be slow. For R packages, the utility included LinkingTo packages when resolving dependencies for binary packages. Consequentially, LinkingTo packages (i.e. BH and Rcpp where many header files exist) can have large files counts, which contribute to longer execution times. Longer because there many more files which need permissions applied when SQL Server uses the launchpad service to instantiate an external script session contained in an AppContainer. 
- launchpad: [Microsoft Docs](https://docs.microsoft.com/sql/machine-learning/concepts/security?view=sql-server-ver15#services-used-in-external-processing-launchpad-1)
- App Container Isolation Process: [Microsoft Docs](https://docs.microsoft.com/sql/machine-learning/install/sql-server-machine-learning-services-2019?view=sql-server-ver15#about-appcontainer-isolation)

## What does this change do?

When the utility calculates package dependencies, it gets type-specific URL paths (binary and source) to the configured CRAN repos and compiles a list of available packages to install. The available binary and source package lists are combined into one list and **are now joined such that the binary package is kept** when a package exists as an entry in both the binary and source lists.
- This is achieved by making `binaryPackages` the first argument in `rbind(binaryPackages, sourcePackages)` so that when the `duplicated()` function runs on the combined list, the first instance of a duplicate (top to bottom) is removed. 
  - rbind() [reference](https://web.mit.edu/~r/current/arch/i386_linux26/lib/R/library/mvbutils/html/rbdf.html)
  - duplicated() [reference](https://www.rdocumentation.org/packages/base/versions/3.5.2/topics/duplicated)

The above change is sufficient in isolation if a user only asks for one package to be installed via `sql_install.packages()`. Now, the utility will iterate over each package requested and properly resolve dependencies (whether to include LinkingTo package dependencies) based on whether the desired package is available as binary, or only as source. 

The function `tools::package_dependencies()` [reference](https://www.rdocumentation.org/packages/tools/versions/3.5.2/topics/package_dependencies) argument `which` includes the LinkingTo packages in the dependency calculation by default ([RDocumentation](https://www.rdocumentation.org/packages/tools/versions/3.5.2/topics/package_dependencies)). The utility now determines whether a package is available as source or binary, and populates the `which` argument accordingly. 

- Binary: `which = c("Depends", "Imports")`
- Source: `which = c("Depends", "Imports", "LinkingTo")`

## How is this change tested?

Two tests are added to validate only the appropriate packages are installed.

1. **Binary Package install with LinkingTo dependency** ensures no LinkingTo packages are installed if the package has an available binary.
2. **Source Package install with LinkingTo dependency** ensures the LinkingTo package dependencies are resolved and are made available for reference when a package is built from source.

### CODE REVIEW

File: R/tests/testthat/test.sqlPackage.dependencies.R [hunk 1]
```diff
@@ -151,3 +151,124 @@ test_that( "Installing a package that is already in use",
...
+
+        #
+        # install the package with its dependencies and validate that the LinkingTo package was not installed
+        #
+        output <- try(capture.output(sql_install.packages( connectionStringAirlineUserdbowner, packageName, verbose = TRUE, scope = scope)))
+        print(output)
+        expect_true(!inherits(output, "try-error"))
+        expect_equal(1, sum(grepl("Successfully installed packages on SQL server", output)))
+
+        helper_checkPackageStatusRequire( connectionStringAirlineUserdbowner,  packageName, TRUE)
```
Comment (Aniruddh25) [inline R/tests/testthat/test.sqlPackage.dependencies.R:196]:
```suggestion
        helper_checkPackageStatusRequire(connectionStringAirlineUserdbowner,  packageName, TRUE)

```

File: R/tests/testthat/test.sqlPackage.dependencies.R [hunk 2]
```diff
@@ -151,3 +151,124 @@ test_that( "Installing a package that is already in use",
...
+        helper_checkPackageStatusRequire( connectionStringAirlineUserdbowner,  linkingToPackageName, FALSE)
```
Comment (Aniruddh25) [inline R/tests/testthat/test.sqlPackage.dependencies.R:197]:
```suggestion
        helper_checkPackageStatusRequire(connectionStringAirlineUserdbowner,  linkingToPackageName, FALSE)

```

File: R/tests/testthat/test.sqlPackage.dependencies.R [hunk 3]
```diff
@@ -151,3 +151,124 @@ test_that( "Installing a package that is already in use",
...
+        packageName <- c("spacefillr")
+        linkingToPackageName <- "Rcpp"
+
+        #
+        # remove old packages if any and verify they aren't there
+        #
+        if (helper_remote.require(connectionStringAirlineUserdbowner, packageName) == TRUE)
+        {
+            cat("\nINFO: removing package:", packageName,"\n")
+            sql_remove.packages( connectionStringAirlineUserdbowner, c(packageName), verbose = TRUE, scope = scope)
```
Comment (Aniruddh25) [inline R/tests/testthat/test.sqlPackage.dependencies.R:236]:
```suggestion
            sql_remove.packages(connectionStringAirlineUserdbowner, c(packageName), verbose = TRUE, scope = scope)

```

File: R/tests/testthat/test.sqlPackage.dependencies.R [hunk 4]
```diff
@@ -151,3 +151,124 @@ test_that( "Installing a package that is already in use",
...
+        }
+
+        if (helper_remote.require(connectionStringAirlineUserdbowner, linkingToPackageName) == TRUE)
+        {
+            cat("\nINFO: removing package:", linkingToPackageName,"\n")
+            sql_remove.packages( connectionStringAirlineUserdbowner, c(linkingToPackageName), verbose = TRUE, scope = scope)
+        }
+
+        helper_checkPackageStatusRequire( connectionStringAirlineUserdbowner, packageName, FALSE)
```
Comment (Aniruddh25) [inline R/tests/testthat/test.sqlPackage.dependencies.R:245]:
```suggestion
        helper_checkPackageStatusRequire(connectionStringAirlineUserdbowner, packageName, FALSE)

```

File: R/tests/testthat/test.sqlPackage.dependencies.R [hunk 5]
```diff
@@ -151,3 +151,124 @@ test_that( "Installing a package that is already in use",
...
+        helper_checkPackageStatusRequire( connectionStringAirlineUserdbowner, linkingToPackageName, FALSE)
```
Comment (Aniruddh25) [inline R/tests/testthat/test.sqlPackage.dependencies.R:246]:
```suggestion
        helper_checkPackageStatusRequire(connectionStringAirlineUserdbowner, linkingToPackageName, FALSE)

```

File: R/tests/testthat/test.sqlPackage.dependencies.R [hunk 6]
```diff
@@ -151,3 +151,124 @@ test_that( "Installing a package that is already in use",
...
+
+        #
+        # install the package with its dependencies and validate that the LinkingTo package was installed
+        #
+        output <- try(capture.output(sql_install.packages( connectionStringAirlineUserdbowner, packageName, verbose = TRUE, scope = scope)))
+        print(output)
+        expect_true(!inherits(output, "try-error"))
+        expect_equal(1, sum(grepl("Successfully installed packages on SQL server", output)))
+
+        helper_checkPackageStatusRequire( connectionStringAirlineUserdbowner,  packageName, TRUE)
```
Comment (Aniruddh25) [inline R/tests/testthat/test.sqlPackage.dependencies.R:256]:
```suggestion
        helper_checkPackageStatusRequire(connectionStringAirlineUserdbowner,  packageName, TRUE)

```

File: R/tests/testthat/test.sqlPackage.dependencies.R [hunk 7]
```diff
@@ -151,3 +151,124 @@ test_that( "Installing a package that is already in use",
...
+        helper_checkPackageStatusRequire( connectionStringAirlineUserdbowner,  linkingToPackageName, TRUE)
```
Comment (Aniruddh25) [inline R/tests/testthat/test.sqlPackage.dependencies.R:257]:
```suggestion
        helper_checkPackageStatusRequire(connectionStringAirlineUserdbowner,  linkingToPackageName, TRUE)

```

**Review by Aniruddh25 — APPROVED** [review]
LGTM!

**Comment by Aniruddh25** [comment]
With this fix, the unneeded packages will no longer be installed so the client scripts which might be depending on those packages  might start to fail without them realizing it. So, I think we should update the sqlmlutils minor version at least and document the new behavior when we release it with this fix.

File: R/R/sqlPackage.R [hunk 8]
```diff
@@ -1155,7 +1155,35 @@ getDependentPackagesToInstall <- function(pkgs, availablePackages, installedPack
...
 
-    dependencies <- tools::package_dependencies(packages = pkgs, db = availablePackages, recursive = TRUE, verbose = FALSE)
+    dependencies <- NULL
+    repos <- getOption("repos")
+    contribWinBinaryUrl <- utils::contrib.url(repos = repos, type = "win.binary")
+
+    #
+    # Build list of dependencies
+    #
+    for ( package in pkgs)
```
Comment (monamaki) [inline R/R/sqlPackage.R:1165]: nit: extra space

**Review by monamaki — APPROVED** [review]

**Review by GarrettBeatty — APPROVED** [review]
Which tests verifies that if there are packages with same name that it chooses binary?

**Review by rashworld-max — APPROVED** [review]

**Pull request merged**